macro/src/class_info.rs

```diff
@@ -424,8 +426,40 @@ impl Type {
         }
     }
 
-    pub fn descriptor(&self) -> String {
-        self.to_non_repeating().descriptor()
+    pub fn descriptor(&self, ctx: &GenericsScope) -> String {
+        self.to_non_repeating().descriptor(ctx)
+    }
+}
+
+/// Track generics currently in scope
+///
+/// In order to resolve descriptors for methods containing `<X extends Y>`, we need to know how `T` was declared.
+pub enum GenericsScope<'a> {
+    Empty,
+    Generics(&'a [Generic], &'a GenericsScope<'a>),
+}
+
+impl<'a> GenericsScope<'a> {
+    /// Find a generic within this scope by name
+    fn find(&self, ty: &Id) -> Option<&Generic> {
+        match self {
+            GenericsScope::Empty => None,
+            GenericsScope::Generics(g, inner) => g.iter().find(|g| &g.id == ty).or(inner.find(ty)),
+        }
+    }
+
+    /// Add an additional layer to this scope (e.g. combining generics from a class with generics from a method)
+    ///
+    /// The newly provided generics are higher priority than the inner generics (but
+    /// I don't think we can have namespace collisions here in Java anyway)
+    fn nest(&'a self, generics: &'a [Generic]) -> GenericsScope<'a> {
+        GenericsScope::Generics(generics, self)
+    }
+}
+
+impl<'a> From<&'a ClassInfo> for GenericsScope<'a> {
```

Comment:
Why not add a method like `class_info.generics_scope()`?

Seems a bit more obvious to me.

Reply:
+1 that's much clearer